java/ql/src/semmle/code/java/security/OgnlInjection.qll

```diff
@@ -0,0 +1,130 @@
+        "com.opensymphony.xwork2.ognl;OgnlUtil;false;setValue;;;Argument[-1..0];ognl-injection",
+        "com.opensymphony.xwork2.ognl;OgnlUtil;false;callMethod;;;Argument[-1..0];ognl-injection"
+      ]
+  }
+}
+
+private class DefaultOgnlInjectionSink extends OgnlInjectionSink {
+  DefaultOgnlInjectionSink() { sinkNode(this, "ognl-injection") }
+}
+
+/** The class `org.apache.commons.ognl.Ognl` or `ognl.Ognl`. */
+private class TypeOgnl extends Class {
+  TypeOgnl() {
+    this.hasQualifiedName("org.apache.commons.ognl", "Ognl") or
+    this.hasQualifiedName("ognl", "Ognl")
+  }
+}
+
+/** The class `org.apache.commons.ognl.Node` or `ognl.Node`. */
+private class TypeNode extends Interface {
+  TypeNode() {
+    this.hasQualifiedName("org.apache.commons.ognl", "Node") or
+    this.hasQualifiedName("ognl", "Node")
+  }
+}
+
+private class TypeExpressionAccessor extends Interface {
+  TypeExpressionAccessor() {
+    this.hasQualifiedName("org.apache.commons.ognl.enhance", "ExpressionAccessor") or
+    this.hasQualifiedName("ognl.enhance", "ExpressionAccessor")
+  }
+}
+
+/**
+ * Holds if `n1` to `n2` is a dataflow step that converts between `String` and `Object` or `Node`,
+ * i.e. `Ognl.parseExpression(tainted)` or `Ognl.compileExpression(tainted)`.
+ */
+private predicate parseCompileExpressionStep(DataFlow::Node n1, DataFlow::Node n2) {
+  exists(MethodAccess ma, Method m, int index |
+    n1.asExpr() = ma.getArgument(index) and
+    n2.asExpr() = ma and
+    ma.getMethod() = m and
+    m.getDeclaringType() instanceof TypeOgnl
+  |
+    m.hasName("parseExpression") and index = 0
+    or
+    m.hasName("compileExpression") and index = 2
+  )
+}
+
+/**
+ * Holds if `n1` to `n2` is a dataflow step that converts between `Node` and `Accessor`,
+ * i.e. `Node.getAccessor()`.
+ */
+private predicate getAccessorStep(DataFlow::Node n1, DataFlow::Node n2) {
+  exists(MethodAccess ma, Method m |
+    n1.asExpr() = ma.getQualifier() and
+    n2.asExpr() = ma and
+    ma.getMethod() = m and
+    m.getDeclaringType().getASupertype*() instanceof TypeNode
+  |
+    m.hasName("getAccessor")
+  )
+}
+
+/**
+ * Holds if `n1` to `n2` is a dataflow step that converts between `Node` and `Accessor`
+ * in a `setExpression` call, i.e. `accessor.setExpression(tainted)`
+ */
+private predicate setExpressionStep(DataFlow::Node n1, DataFlow::Node n2) {
+  exists(MethodAccess ma, Method m |
+    n1.asExpr() = ma.getArgument(0) and
+    n2.asExpr() = ma.getQualifier() and
+    ma.getMethod() = m and
+    m.getDeclaringType().getASupertype*() instanceof TypeExpressionAccessor
+  |
+    m.hasName("setExpression")
+  )
+}
+
+private class DefaultOgnlInjectionAdditionalTaintStep extends OgnlInjectionAdditionalTaintStep {
+  override predicate step(DataFlow::Node node1, DataFlow::Node node2) {
+    parseCompileExpressionStep(node1, node2) or
+    getAccessorStep(node1, node2) or
+    setExpressionStep(node1, node2)
+  }
```

Comment:
I was going to suggest that this could be done using CSV rows with a custom `kind` column, but it doesn't look like `ExternalFlow` exposes a predicate for that. 

@aschackmull would it make sense to use CSV for the additional flow steps of queries like this?

Reply:
I thought of doing that too, for consistency. But apparently adding CSV summaries in queries is currently not supported/encouraged because it has an impact on performance -- that's why the predicate isn't exposed.